feat: add `string/base/slice`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds `@stdlib/string/base/slice` function including documentation, benchmarks and tests.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/33

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add string/base/slice

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode committed:
test: add test case

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/string/base/slice/test/test.js:
```diff
@@ -0,0 +1,120 @@
+	t.end();
+});
+
+tape( 'the function supports slicing the UTF-16 code units of a provided string (positive indices)', function test( t ) {
```
```suggestion
tape( 'the function slices the UTF-16 code units of a provided string (positive indices)', function test( t ) {
```

headlessNode commented on lib/node_modules/@stdlib/string/base/slice/test/test.js:
```diff
@@ -0,0 +1,120 @@
+	t.end();
+});
+
+tape( 'the function slicing the UTF-16 code units of a provided string (negative start index)', function test( t ) {
```
```suggestion
tape( 'the function slices the UTF-16 code units of a provided string (negative start index)', function test( t ) {
```

headlessNode commented on lib/node_modules/@stdlib/string/base/slice/test/test.js:
```diff
@@ -0,0 +1,120 @@
+	t.end();
+});
+
+tape( 'the function slicing the UTF-16 code units of a provided string (negative end index)', function test( t ) {
```
```suggestion
tape( 'the function slices the UTF-16 code units of a provided string (negative end index)', function test( t ) {
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/string/base/slice/docs/repl.txt:
```diff
@@ -0,0 +1,31 @@
+        Input string.
+
+    start: integer
+        Index for the first character to include.
```
```suggestion
        Slice start index (inclusive).
```

headlessNode commented on lib/node_modules/@stdlib/string/base/slice/docs/repl.txt:
```diff
@@ -0,0 +1,31 @@
+        Index for the first character to include.
+
+    end: integer
+        Index for the first character to exclude.
```
```suggestion
        Slice end index (exclusive).
```

headlessNode commented on lib/node_modules/@stdlib/string/base/slice/docs/types/index.d.ts:
```diff
@@ -0,0 +1,58 @@
+* Slices UTF-16 code units from a string.
+*
+* @param str - input string
+* @param start - index for first character to include
+* @param end - index for first character to exclude
```
```suggestion
* @param start - slice start index (inclusive)
* @param end - slice end index (exclusive)
```

headlessNode commented on lib/node_modules/@stdlib/string/base/slice/lib/main.js:
```diff
@@ -0,0 +1,49 @@
+* Slices UTF-16 code units from a string.
+*
+* @param {string} str - input string
+* @param {integer} start - index for the first character to include
+* @param {integer} end - index of the first character to exclude
```
```suggestion
* @param {integer} start - slice start index (inclusive)
* @param {integer} end - slice end index (exclusive)
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/string/base/slice/README.md:
````diff
@@ -0,0 +1,96 @@
+out = slice( 'foo bar', -1, 7 );
+// returns 'r'
+```
+
````
@headlessNode For clarity, would you also mind documenting the parameters here? In particular, it is not immediately obvious without studying the code that `start` is inclusive and `end` exclusive.

kgryte commented on lib/node_modules/@stdlib/string/base/slice/docs/types/test.ts:
```diff
@@ -0,0 +1,68 @@
+	slice( 'abc', 1, 4 ); // $ExpectType string
+}
+
+// The compiler throws an error if the function is provided a value other than a string...
```
```suggestion
// The compiler throws an error if the function is provided a first argument that is not a string...
```

kgryte committed:
docs: update comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/string/base/slice/lib/main.js:
```diff
@@ -0,0 +1,49 @@
+* @param {string} str - input string
+* @param {integer} start - slice start index (inclusive)
+* @param {integer} end - slice end index (exclusive)
+* @throws {TypeError} function must be provided three arguments
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/string/base/slice/lib/main.js:
```diff
@@ -0,0 +1,49 @@
+function slice( str, start, end ) {
+	if ( arguments.length !== 3 ) {
+		throw new TypeError( 'invalid invocation. Function must be provided a string, starting index and an ending index.' );
+	}
```
```suggestion
```

"base" implementations can make additional assumptions, including that all arguments are provided.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: cache reference to built-in

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/string/base/slice/test/test.js:
```diff
@@ -0,0 +1,120 @@
+	t.end();
+});
+
+tape( 'the function throws an error if provided an unsupported number of arguments', function test( t ) {
+	var values;
+	var i;
+
+	values = [
+		[ 'foo', 1 ],
+		[ 'bar' ],
+		[ 'beep', 1, 3, 5 ]
+	];
+
+	for ( i = 0; i < values.length; i++ ) {
+		t.throws( badValue.apply( null, values[i] ), TypeError, 'throws an error when provided unsupported number of arguments' );
+	}
+	t.end();
+
+	function badValue() {
+		return function badValue() {
+			slice.apply( null, arguments );
+		};
+	}
+});
+
```
```suggestion
```

kgryte committed:
test: remove test

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (commented):
This looks good. Just a small nit to document arguments.

kgryte commented on lib/node_modules/@stdlib/string/base/slice/lib/main.js:
```diff
@@ -0,0 +1,50 @@
+// Cache a reference to the built-in to prevent prototype mutation shenanigans:
+var sliceString = String.prototype.slice;
+
+// MAIN //
```
```suggestion

// MAIN //
```

kgryte committed:
style: add empty line

Signed-off-by: Athan <kgryte@gmail.com>

headlessNode committed:
docs: add function args description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/string/base/slice/README.md:
```diff
@@ -0,0 +1,102 @@
+
+-   **str**: input string.
+-   **start**: slice start index (inclusive).
+-   **end**: slice eng index (exclusive).
```
```suggestion
-   **end**: slice end index (exclusive).
```

headlessNode committed:
docs: fix typo

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte reviewed (approved):
LGTM
